File: offer/序列化二叉树.cpp
/*
请实现两个函数，分别用来序列化和反序列化二叉树
*/

/*
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};
*/
class Solution {
public:
    char* Serialize(TreeNode *root) {
        stringstream ss;
        serialize(root, ss);
        string str = ss.str();
        char* res = new char[str.length()+1];
        strcpy(res, str.c_str());
        return res;
    }
    void serialize(TreeNode *root, stringstream &ss) {
        if ( !root ) {
            ss << "#";
            return;
        }
        ss << to_string(root->val);
        serialize(root->left, ss);
        serialize(root->right, ss);
    }
    TreeNode* Deserialize(char *str) {
        stringstream ss;
        ss << str;
        TreeNode* root = deserialize(ss);
        return root;
    }
    TreeNode* deserialize(stringstream &ss) {
        string str;
        ss >> str;
        if ( str == "#" )
            return NULL;
        TreeNode* root = new TreeNode(atoi(str.c_str()));
        root->left = deserialize(ss);
        root->right = deserialize(ss);
        return root;
    }
};

File: offer/序列化二叉树.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def __init__(self):
        self.index = -1

    def Serialize(self, root):
        if not root:
            return "#."
        return str(root.val)+"." + self.Serialize(root.left) + self.Serialize(root.right)

    def Deserialize(self, s):
        self.index = -1
        vals = s.split(".")[:-1]
        print(vals)
        return self.deserialize(vals)

    def deserialize(self, vals):
        self.index += 1
        if self.index >= len(vals):
            return None
        if vals[self.index] == "#":
            return None
        root = TreeNode(int(vals[self.index]))
        root.left = self.deserialize(vals)
        root.right = self.deserialize(vals)
        return root


if __name__ == '__main__':
    s = Solution()
    s.Deserialize("1.2.3.")
